proof/refine/RISCV64/Invariants_H.thy

```diff
@@ -130,29 +167,151 @@ lemma st_tcb_at'_def:
+
+abbreviation replySCs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "replySCs_of s \<equiv> replies_of' s |> replySC"
+
+abbreviation sc_of' :: "kernel_object \<Rightarrow> sched_context option" where
+  "sc_of' \<equiv> projectKO_opt"
+
+abbreviation scs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> sched_context option" where
+  "scs_of' s \<equiv> ksPSpace s |> sc_of'"
+
+abbreviation scReplies_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "scReplies_of s \<equiv> scs_of' s |> scReply"
+
+abbreviation tcb_of' :: "kernel_object \<Rightarrow> tcb option" where
+  "tcb_of' \<equiv> projectKO_opt"
+
+abbreviation tcbs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> tcb option" where
+  "tcbs_of' s \<equiv> ksPSpace s |> tcb_of'"
+
+abbreviation tcbSCs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "tcbSCs_of s \<equiv> tcbs_of' s |> tcbSchedContext"
+
+abbreviation scTCBs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "scTCBs_of s \<equiv> scs_of' s |> scTCB"
+
+abbreviation sym_heap_tcbSCs where
+  "sym_heap_tcbSCs s \<equiv> sym_heap (tcbSCs_of s) (scTCBs_of s)"
+
+abbreviation sym_heap_scReplies where
+  "sym_heap_scReplies s \<equiv> sym_heap (scReplies_of s) (replySCs_of s)"
+
 definition tcb_cte_cases :: "machine_word \<rightharpoonup> ((tcb \<Rightarrow> cte) \<times> ((cte \<Rightarrow> cte) \<Rightarrow> tcb \<Rightarrow> tcb))" where
  "tcb_cte_cases \<equiv> [ 0 << cteSizeBits \<mapsto> (tcbCTable, tcbCTable_update),
                     1 << cteSizeBits \<mapsto> (tcbVTable, tcbVTable_update),
-                    2 << cteSizeBits \<mapsto> (tcbReply, tcbReply_update),
-                    3 << cteSizeBits \<mapsto> (tcbCaller, tcbCaller_update),
-                    4 << cteSizeBits \<mapsto> (tcbIPCBufferFrame, tcbIPCBufferFrame_update) ]"
+                    2 << cteSizeBits \<mapsto> (tcbIPCBufferFrame, tcbIPCBufferFrame_update),
+                    3 << cteSizeBits \<mapsto> (tcbFaultHandler, tcbFaultHandler_update),
+                    4 << cteSizeBits \<mapsto> (tcbTimeoutHandler, tcbTimeoutHandler_update) ]"
 
 definition max_ipc_words :: machine_word where
   "max_ipc_words \<equiv> capTransferDataSize + msgMaxLength + msgMaxExtraCaps + 2"
 
-definition tcb_st_refs_of' :: "thread_state \<Rightarrow> (obj_ref \<times> reftype) set" where
-  "tcb_st_refs_of' st \<equiv> case st of
-     (BlockedOnReceive x _)     => {(x, TCBBlockedRecv)}
-   | (BlockedOnSend x _ _ _ _)  => {(x, TCBBlockedSend)}
-   | (BlockedOnNotification x)  => {(x, TCBSignal)}
-   | _                          => {}"
+type_synonym ref_set = "(obj_ref \<times> reftype) set"
+
+definition tcb_st_refs_of' :: "thread_state \<Rightarrow> ref_set" where
+  "tcb_st_refs_of' z \<equiv> case z of
+    Running                    => {}
+  | Inactive                   => {}
+  | Restart                    => {}
```

Comment:
should use a default pattern `_ => {}` for these cases, so the the definition is resistant to change in that datatype